### ISSUE TITLE
Follow a route

### ISSUE THREAD

**Issue by HarelM** [IsraelHikingMap/Site#929]
## Feature
As a user I would like to be able to define a route to follow.

This is a placeholder feature for the discussion of how to support the concept of following a route.

**Comment by HarelM** [comment]
Part of it might be #838 - display a sound when getting lost.

**Comment by HarelM** [comment]
From #1134 as @zstadler wrote:
Some features that could be valuable in this mode:

Cross-Track alarm: when the GPS position has diverged from the planned route more than a given value.
Route-based Statistics: remaining distance to target, estimated time to target, etc.
Re-route based on GPS position.
Provide directions

**Comment by valleyofdawn** [comment]
**Feature**: identify location on a planned route
**Purpose**: To identify your location within a planned route, assess time to completion and prepare for ascents and descents to come.
In the future, some of the code can be used for other features, such as follow route, off-route alarm, etc.
**Means**:
Given that there is an active route and current location, calculate whether the current location is within less than 50 m (horizontally only) of the route, and if so, mark the nearest point on route height profile.
**Performance issues:**
Can be activated only when the height profile is shown
Can be updated every 10 seconds or so.
Once a match is identified, the active route can be scanned only within a few points befor and after the previous point.
**Appearance**: Three possibilities shown:
![image](https://user-images.githubusercontent.com/689201/74652985-e73e1480-518f-11ea-919d-258ab6334e40.png)

**Comment by HarelM** [comment]
This does not take into consideration what happens when recording and does not solve how much time is left (unless you meant distance and not time).
While the implementation of this is fairly simple I'm not sure it solves much and I would like to avoid writing this feature multiple times.
I suggest to try and formulate the design concept in a phone conversation.

**Comment by valleyofdawn** [comment]
I thought that can be a first stage, before dealing with recording.
I think we should leave the assessment of time to finish to the user. This is just a tool to do so. Such predictions are inaccurate when the terrain is variable.
If on the other hand you want to deal with recording on the first go than we should add a "follow this route" button and not rely on the active route. In such a case the followed route height profile should replace the recorded route and we should add an extra line to the stats table (maybe distance to go, ETA, current elevation?).
To be discussed over the phone.

**Comment by valleyofdawn** [comment]
possible icons
![image](https://user-images.githubusercontent.com/689201/74661130-06dd3900-51a0-11ea-8b78-1824120994fa.png)

**Comment by valleyofdawn** [comment]
A suggestion on how the stats and height profile screens can look under different situations.
![Follow route](https://user-images.githubusercontent.com/689201/75276696-56121200-580f-11ea-83a0-7b21624e41ba.png)

**Comment by valleyofdawn** [comment]
As for selecting a route to follow, we can adopt the format for format waypoint when route editing is off.
![follow selection](https://user-images.githubusercontent.com/689201/75278533-b8204680-5812-11ea-93be-c4323e4a04a3.png)

**Comment by HarelM** [comment]
As discussed on the phone - there's no need select a route - a route will be selected automatically if one exists and is close enough. If this wouldn't be enough we'll rethink it.
Also please send a definition of what should the lower bar look like when it is not open in all the cases above.

**Comment by HarelM** [comment]
Another issue which we need to define is the problem between planned route and recorded route in terms of distance and height:
When recording, the distance traveled is different than the planned route distance, I see this a lot of times when hiking.
The suggestion above draws the recorded route on top of the planned route in the height profile.
This height profile x axis is the traveled distance, What should the user see when recording a route while the height profile is open in terms of distance? the planned route scaled to match the recorded route or the recorded route scaled to match the planned route?
Same question for the height itself.

I believe the solution should be that when recording and following a route the recorded route in the height profile should be scaled to the planned route, i.e. only "mark" the points on the planned route which are close enough.

Another thing to consider is when starting recording before a planned route started and when continue recording after the planned route ended.
Options are:
1. Ignore those points and only show points that match in case there are any (if there are non, then the recording will behave as there's no planned route)
2. When current recorded point is not close enough to a planned route, show a regular recorded route
3. Both of the above - when not close enough show recording without planned, when close enough mark the the points, when going beyond the existing planned route return to show recording without planned.
4. (I think this is complicated to implement, so this is my least favorite option) In case the recorded started before the planned route started add the recorded points to the planned route at the begging (this will happen only when getting close enough to a planned route).
In case the recorded route continued after the planned route ended the recorded points should be added at the end after the planned route.

**Comment by valleyofdawn** [comment]
1. For the reasons you mentioned I don’t think the recording should be drawn on top of the planned route. Instead, the data up to the current position should be taken from the recording, and the data from the current position onward – taken from the planned route. Let’s call it the “hybrid route”.

2. The X and Y axes should reflect the maximum range of the hybrid route. That way, you’d know at a glance what proportion of the distance and climb you already covered. If on the other hand, you decided to climb a hill in the middle of the flat route you had planned, it won’t be cropped from the height profile.

3. In the future we can consider zooming in and out of the profile, but this is an overkill at this stage

4. Regarding deviations from the planned route, be them before, during, or after following the route, I think we should set a minimum distance (say 100m) and stop showing the planned route if it is exceeded. You can resume showing it when you are back on track.

**Comment by HarelM** [comment]
1. Is basically option 4 from my comment, right? If so I would consider this as an enhancement, since options 1,2 and 3 are far simple I think we should focus on the important part of this feature. 
2. Is not clear to me, please add a screenshot/mock up.
3. Zooming is an overkill, I don't see a valid use case for this - if your watching the chart when riding you don't have time to play with it, and there's little interesting data in it anyway.
4. I though about 50 meters, but your call...

**Comment by zstadler** [comment]
@valleyofdawn wrote:
> 1. For the reasons you mentioned I don’t think the recording should be drawn on top of the planned route. Instead, the data up to the current position should be taken from the recording, and the data from the current position onward – taken from the planned route. Let’s call it the “hybrid route”.

I like this approach, despite its technical complexity, because all divergences from the original plan are fully considered in the combined view. It is a modified version of 

@HarelM wrote:
> 4. (I think this is complicated to implement, so this is my least favorite option) In case the recorded started before the planned route started add the recorded points to the planned route at the begging (this will happen only when getting close enough to a planned route).
In case the recorded route continued after the planned route ended the recorded points should be added at the end after the planned route.

where the head of planned route until the current position is removed and the entire recording is added before it.

**Comment by zstadler** [comment]
@valleyofdawn wrote:
> 4. Regarding deviations from the planned route, be them before, during, or after following the route, I think we should set a minimum distance (say 100m) and stop showing the planned route if it is exceeded. You can resume showing it when you are back on track.

I think keeping the planned route on the map is important especially when a diversion happens. We cannot assume that every diversion is voluntary, conscience, or limited in distance.

Showing the planned route will help the user to get back on track. If the planned route is no longer valuable to the user, the user can choose to stop the follow-route mode.

**Comment by HarelM** [comment]
There are no plans to change the route on the map, only the height graph.

**Comment by zstadler** [comment]
:+1:

**Comment by valleyofdawn** [comment]
So to explain what happens to the axes and when you deviate from the planned course:
![image](https://user-images.githubusercontent.com/689201/81094236-e2c20600-8f0b-11ea-80ba-54a691722dc2.png)

**Comment by valleyofdawn** [comment]
I like your idea of not actually choosing a route to follow.
What did you have in mind for the following scenarios, though?
1) You have two variations of the route, which one is followed?
2) what happens when your track crossed a route on the map.
3) what happens if you follow a route opposite its direction?

**Comment by HarelM** [comment]
1. Always the closest one - I don't think this is a popular use case, not many people use more than one route.
2. If there's no other planned route, momentarily you'll see this route in the height graph until you are far enough
3. That's what's nice about my suggestion (option 3) - it will mark the coordinates/segments which are close enough to the recorded route, direction doesn't matter.
Again, I'm only talking about the height graph, not the route on the map and probably not the statistics.
Basically coloring the planned route "closest points" or showing the recorded route itself if there's no close enough planned route - it is a simple solution to implement and doesn't have tons of edge cases to define and maintain.

**Comment by zstadler** [comment]
What is the expected behavior when the planned route crosses itself, as described in https://github.com/IsraelHikingMap/Site/issues/1203?

**Comment by HarelM** [comment]
I have just pushed an initial commit for this feature - this is **not** the final version.
When not in hybrid mode it's fairly simple - only show the position of the marker on the graph.
I haven't touched the statistics part as I'm still waiting for a response to the following question:
>Also please send a definition of what should the lower bar look like when it is not open in all the cases above.

In the current version when recording the graph will switch to the closest planned route to the current GPS position.
I know, this is not ideal and can be greatly improved but I would like to get an initial feeling of how this works before spending too much time one this.

Still planned for the initial version:
1. Coloring the already traveled planned route in case of hybrid mode only
2. Improve statistics - pending the requested definition above, this should take into consideration the case of following the route in the opposite direction, although I think that in this case the user should just switch the planned route direction instead of trying to figure this out in the code and complicating everything...

@valleyofdawn Please also send the svg of the icons.

**Comment by valleyofdawn** [comment]
1) Cool, let me know when there's a beta I could test
2) Regarding a looping planned route. I don'e see a problem. I presume there must be an algorithm that identifies which point in the planned route is closest to the current position. In that case you'd go down the correct side of the split. There might be a second of indecision, but once you move away in one direction it should be fine.
3) The minimal display when the stats/profile window is closed could be identical to the top row of the table in all 4 cases.

**Comment by HarelM** [comment]
1- I have uploaded the version for internal testers, feel free to play with it.
3- The only problem I have with it is that remaining time will now be shown in half of the cases - I don't see a good way to make it reliable - please elaborate on it's formula.
A possible solution, which is mediocre, is to divide remaining distance in current speed or average speed if recording. Bare in mind that when standing still and not recording, the speed is "-" and I guess the remaining time will be "--:--".
We can also use ETA instead of remaining time, i.e. the hour which you are planned to finish the route.
If you have another idea, please share.
I still need the icons...

**Comment by valleyofdawn** [comment]
I think ETA is more useful than remaining time.
Calculating based on average speed is better than current speed (which is erratic). Average of last 15 minutes is also an option.
Seems like GraphHopper has an ETA function. Perhaps its more advanced than that. Can we use it?

**Comment by HarelM** [comment]
Initial tests I made today, all-in-all works as expected, I see very little use to this feature as a hiker.
1. There's a need for a vertical line on the location and see the distance on the x axis.
2. There's a bug in the color of the route, the planned route looks orange instead of staying with its color when GPS is close enough.

We can't use grasshopper since we might not be connected to the server, also I doubt it'll give good results. I'll go with average if exists or current speed if not.

**Comment by valleyofdawn** [comment]
1. Works well for me too. 50m is the cut-off? Seems better than the 100m I suggested.
2. Does it only refer to the active route right now? I couldn't get it to auto-switch planned routes. Will try again.
3. This would be very useful for bikers, and even this implementation gives you 80% of the functionality.
4. Here's a suggestion for the height and distance indicators.
 ![image](https://user-images.githubusercontent.com/689201/81263160-db077c00-9047-11ea-8a2d-aaee43496d8d.png)
5. I think another quick-fix now is to add an option for off-route alert in the advanced settings. Simply buzz when you stop showing the position indicator on the  profile and beep when it is regained.


[ETA.svg.txt](https://github.com/IsraelHikingMap/Site/files/4591242/ETA.svg.txt)
[remaining.svg.txt](https://github.com/IsraelHikingMap/Site/files/4591244/remaining.svg.txt)
[travelled.svg.txt](https://github.com/IsraelHikingMap/Site/files/4591245/travelled.svg.txt)

**Comment by HarelM** [comment]
1- Yes, 50m is the cut off currently, can be changed easily. It is the cut-off for merge routes as well, I think it's a good value.
2- You are right, only when recording the planned route will change, not when only GPS is active, I guess I'll change this behavior to find the closest non-recording route when GPS is active. good catch!
4- I have implemented vertical line, let's see if this is good enough, the provided picture is a bit tricky to implement.
5- not a quick fix, but part of this issue in general, although I might move it to another issue since it's big.
6- Thanks!

**Comment by HarelM** [comment]
Icons are too small :-( See if you can create a square icon that will take the entire canvas.
Planning on using ETA instead of an icon as it is more common.
Below are some screen shots:
![image](https://user-images.githubusercontent.com/3269297/81275047-ca600180-9059-11ea-9cd3-54c72c3f59dd.png)
![image](https://user-images.githubusercontent.com/3269297/81275079-d77cf080-9059-11ea-9340-910f50793235.png)
![image](https://user-images.githubusercontent.com/3269297/81275109-e19eef00-9059-11ea-8e82-211cdb23ed97.png)
![image](https://user-images.githubusercontent.com/3269297/81275143-ec598400-9059-11ea-9cee-06f2893a98ff.png)
![image](https://user-images.githubusercontent.com/3269297/81275300-2591f400-905a-11ea-9c78-877a0e8884d9.png)

**Comment by valleyofdawn** [comment]
The SVGs as you asked
[remaining.svg.txt](https://github.com/IsraelHikingMap/Site/files/4592430/remaining.svg.txt)
[travelled.svg.txt](https://github.com/IsraelHikingMap/Site/files/4592431/travelled.svg.txt)

**Comment by valleyofdawn** [comment]
I think the ones you show take up the same widths as the existing icons, so the position within the canvas is not the root of the problem.
I can create shorter and bolder arrows if you like.

**Comment by HarelM** [comment]
I decided to use an exiting icon I had, and keep the traveled distance the same as today.
![image](https://user-images.githubusercontent.com/3269297/81287412-61818500-906b-11ea-9572-4d4aa6fc855d.png)
Latest version was just released to internal testing has most of the feature requirements here.
Editing multiple routes while the GPS is on and the graph is open might be confusing due to latest changes. I think it's a rare case and the UX of this feature is more important than this case.
I'm not sure there's a need to color the height graph based on the complexity, helpfulness and reliability - the map itself has the planned and recorded route visible which is more accurate than the height graph.
I'm leaning towards closing this issue assuming everything that was implemented works as expected and opening another issue related to off-route alert.

**Comment by valleyofdawn** [comment]
Please reconsider, there is nothing in the two color-inverted icons to tell the user if its distance-to or distance-from

**Comment by HarelM** [comment]
Try it out before you judge it :-) Actually you and I had the exact opposite understanding of the icons you published - I understood the circle and then arrow as the traveled distance - i.e. circle indicate the start point, and the other one as remaining distance - i.e the circle is the end point.
Using the same icon the users know to the traveled distance and using a negative for remaining makes a lot of sense to me while letting the users keep their current understanding of the UI.

I have tested it to be working pretty good. The ETA was on the spot for a short walk I did with my dogs, there was a bug in the average speed calculation I just fixed, make sure you use the latest version.
This is a nice addition, I admit I was wrong to underestimate it.

**Comment by HarelM** [comment]
Morning test:
Average speed is working as expected.
The feature is working nicely all-in-all.
I've seen the following behavior to be confusing, not sure I want to fix it though:
1. Plan a route
2. Start recording
3. See that I'm going not according to planned route
4. Go to Layers
5. Switch to change the selected route to be the planned and not the recording
6. Edit the planned route to be according to where I'm headed
7. Look at the statistics
--> The distance traveled in this case will be according to the location of the GPS on the planned route and not the recording distance I traveled
A simple workaround is to switch the active route to be the recording again and not the planned.
Not sure how interesting this scenario is though...

**Comment by valleyofdawn** [comment]
Regarding the traveled vs. to-go icons you got it right of course, I got the labels swapped.
I've had multiple issued during my weekend ride. Some may be related to the new features, some not. It's not well documented so I'll try to reconstruct them one-by-one and report. 
Anyway, please don't rush it to production before we thoroughly debug them.

**Comment by HarelM** [comment]
Thanks for taking the time and test this feature!
Make sure to use the latest version from the store when trying to reproduce - I have just uploaded it for internal testing (there was a small issue when turning of the GPS).
I don't think I changed anything besides the route statistics, other bug might already have existed regardless of this feature so I won't consider them as showstoppers for production.

**Comment by valleyofdawn** [comment]
First issue:
The fix of the recording to the planned route was lost more than 10 times during the session even though the current position was spot on the planned route on the display.
You can examine the log I emailed.
This is the share of plan and recording: 
[](https://israelhiking.osm.org.il/share/7NdT863tbs)
A screenshot of such an instance:
![image](https://user-images.githubusercontent.com/689201/81537055-436e9a00-9375-11ea-822e-a13fc9a0bf71.png)

**Comment by valleyofdawn** [comment]
When a planned route is circular, at the starting position the current position "fix" sometimes goes to the end of the route rather than to its start. Similarly at the end of the recording it may fix to the start of the plan route.
When the routes folds over itself like so:
![image](https://user-images.githubusercontent.com/689201/81537612-238ba600-9376-11ea-9bd5-7b3deb6e997a.png)
then the fix keeps skipping between the two relevant sections of the planned route.
An umproved algorithm would look not only at the nearest point, but also at the sequence of waypoints along the planned route.
Two such examples:
![image](https://user-images.githubusercontent.com/689201/81537870-85e4a680-9376-11ea-99f5-eb70582ee6cc.png)
![image](https://user-images.githubusercontent.com/689201/81537907-972db300-9376-11ea-98b6-f9a3859086c4.png)

**Comment by HarelM** [comment]
I failing to follow.
Please open a separate issue on each different bug.
In each case please provide the time date, and relevant log file, preferably only the relevant log lines: +- 5 minutes from the event.

**Comment by valleyofdawn** [comment]
I think a better order for the parameters in the table would be traveled distance on the left, then distance to go, then ETA. It works better with the direction of the height graph and follows logically. 
I tried to get used to the icons and I still find them too hard to distinguish. Attaching the SVGs I suggested, with the right labeling, please reconsider. 
[TO GO.svg.txt](https://github.com/IsraelHikingMap/Site/files/4608316/TO.GO.svg.txt)
[travelled.svg.txt](https://github.com/IsraelHikingMap/Site/files/4608317/travelled.svg.txt)

**Comment by HarelM** [comment]
I've tested the new icon, they still look too small and still confusing.
I'm not sure switching the direction of the bar is better than current one as the ETA kinda makes you read the line and think it is in English, causing the icon to be at the start and not at the end:
![image](https://user-images.githubusercontent.com/3269297/81606140-32f10a80-93db-11ea-8de6-9901f69c7671.png)
I don't have hard opinion on this, I'll change it so you can test it.
